gen_breaktest.go

```diff
@@ -0,0 +1,215 @@
+// +build generate
```

Comment:
Is it common to put "go generate" files in the same directory as the package itself? My IDE gives me warnings here but it may just be a configuration problem.

What happens when we put this file into a subdirectory, e.g. `testcases/generate.go`? I guess then we can't run "go generate" in the main directory anymore, right?

Also, what are the naming conventions for these files? `gen_...` as used here or `generate_...`?

Reply:
> Is it common to put "go generate" files in the same directory

Yes it is, you can see examples of this in the Go standard packages dating way back. Personally, I find it convenient because I can do a simple `go generate && go test` without paths and also paths don't need be listed on the `//go:generate` lines or elsewhere.

> My IDE gives me warnings here but it may just be a configuration problem.

Tools should recognise the build constraint and (mostly) ignore the file. The only thing I can think of is if the tool special cases `+build ignore` but not `+build generate` and assumes that the package should be build-able with `go build -tags generate` (it isn't, and shouldn't be). The `ignore` tag is the canonical, "this file is never a part of this package" and used to be what I used for these, but then you need to put the `//go:generate` comment in some other file. In [Go1.13](https://golang.org/doc/go1.13#go-command) they made it so that `go generate` itself understands and sets the `generate` build tag which makes it possible to put the `//go:generate` comment in the file itself which I find preferable for stand-alone Go generate commands such as this one.

> Also, what are the naming conventions for these files

I'm not aware of any conventions as-such. Based on the Go source itself, it varies.

For plain `gen.go` there is:

- `index/suffixarray/gen.go`
- `image/internal/imageutil/gen.go`
- `image/color/palette/gen.go` (with basically nothing but a comment and `//go:generate …` in a separate `generate.go` file, this was before the `generate` build tag was added to go generate)
- `crypto/md5/gen.go`

For `generate_…` there is:

-  `crypto/tls/generate_cert.go` (not used by any //go:generate I saw)
- `src/time/tzdata/generate_zipdata.go`

For `gen…` there is:

- `sort/genzfunc.go`
- `time/genzabbrs.go`
- `cmd/go/internal/test/genflags.go`
- `hash/crc32/gen_const_ppc64le.go`
- `runtime/internal/sys/gengoos.go`
- `math/rand/gen_cooked.go`

In this case I just picked the "breaktest.go" suffix without too much thought other than it seemed to describe it reasonably well. An argument could also be made for naming the file that generates `foo_test.go` `foo_gen.go`. (I've never actually used go generate to generate test source files before, only regular package source files).

Reply:
Thank you for your explanations. It all makes sense. Let's keep it this way.

Btw, the warning I'm getting in VS Code is this (via `go list`):

> No packages found for open file uniseg/gen_breaktest.go: <nil>.
> If this file contains build tags, try adding "-tags=<build tag>" to your gopls "buildFlags" configuration (see (https://github.com/golang/tools/blob/master/gopls/doc/settings.md#buildflags-string).
> Otherwise, see the troubleshooting guidelines for help investigating (https://github.com/golang/tools/blob/master/gopls/doc/troubleshooting.md).

It appears right here:

[`package main`](https://github.com/dchapes/uniseg/blob/3cc8b8d61c5ecdaec23269f470b15f1118696b3f/gen_breaktest.go#L9) in `gen_breaktest.go`

I guess I need to set the "-tags" parameter now?